src/Symfony/Component/Messenger/Transport/InMemoryTransport.php

```diff
@@ -58,21 +67,28 @@ public function ack(Envelope $envelope): void
     public function reject(Envelope $envelope): void
     {
         $this->rejected[] = $envelope;
+        $id = $this->findMessageIdStamp($envelope)->getId();
+        unset($this->sent[$id]);
```

Comment:
To me it looks like the rejection and acknoledgement did not remove the sent messages on purpose, so that in tests you can call `get` to test if all expected messages have been sent.
To keep this use-case and solve the infinite receiving, I'd propose to have to different properties: 
- `$this->sent` keeps all sent messages and we and we add a getter `getSent()` for them
- `$this->queue` also keeps the sent messages but they get removed on `get()` like you proposed here